enhancements/network/service-egress-traffic-steering.md

```diff
@@ -0,0 +1,301 @@
+                   │                            │               │                │
+                   └────────────────────────────┘               └────────────────┘
+                3. from the mgmt port it hits ovn-worker's
+                   routes and iptables rules
+```
+As mentioned earlier, for the opposite direction (ingress/external client initiates) to work properly the LoadBalancer provider needs to announce the service only from `ovn-worker`.
+
+### Node Selection
+Selecting a node will work similarly to how EgressIP selects one.
+When a LoadBalancer service is annotated with a valid `k8s.ovn.org/egress-service` annotation, a node is selected to be in charge of all of its traffic.
+If the `nodeSelector` field is specified, only a node whose labels match the specified selectors can be selected.
+A cache of the nodes and their number of allocations is kept in order to spread the allocations between all of the nodes available for a given service - selecting the node with the least amount of allocations every time.
+
+Once a node is selected, it is checked for readiness to serve traffic every x seconds the same way EgressIP does for its nodes (TCP/gRPC).
+If a node fails the health check, its allocations move to another node by removing the `egress-service.k8s.ovn.org/<svc-namespace>-<svc-name>: ""` label from it, removing the logical router policies from the cluster router and resetting the `k8s.ovn.org/egress-service-host=<node_name>` annotation on the service which causes it to be reconciled - causing a new node to be selected for the service.
+If the node becomes not ready or its labels no longer match the service's selectors the same re-election process happens.
+
+### Usage Example
+
+With all of the above implemented, a user can follow these steps to create a functioning LoadBalancer service whose endpoints exit the cluster with its IP using MetalLB.
+
+1. Create the IPAddressPool with the desired IP for the service. It makes sense to set `autoAssign: false` so it is not taken by another service - our service will request that pool explicitly. 
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: IPAddressPool
+metadata:
+  name: example-pool
+  namespace: metallb-system
+spec:
+  addresses:
+  - 172.19.0.100/32
+  autoAssign: false
+```
+
+2. Create the LoadBalancer service. We create it with 2 annotations:
+- `metallb.universe.tf/address-pool` - to explicitly request the IP to be from the `example-pool`.
+- `k8s.ovn.org/egress-service` - to request that all of the endpoints of the service exit the cluster with the service's IP. We also provide `nodeSelector` so that the traffic exits from a node that matches these selectors.
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: example-service
+  namespace: some-namespace
+  annotations:
+    metallb.universe.tf/address-pool: example-pool
+    k8s.ovn.org/egress-service: '{"nodeSelector":{"matchLabels":{"color": "green"}}}'
+spec:
```

Comment:
you are missing the ```externalIPs``` field here right? The SNAT'ed IP will only be to the external IP field of a service. Also, externalIPs can have multiple IP addresses. So do we just only use the first one in the list? What about dualstack?

Reply:
I think since we're aiming for loadbalancer services we will use the status.loadbalancer.ingress which contains at most one ip of each family(?)
then according to the status we'll create the relevant policies/snats (covering dual stack if both families appear)
afaik we don't need to cover the externalips case

Reply:
lets clarify in the enhancement we are targeting "ingress IP" and not "external IP"

Reply:
done